marquez_client/client.py

```diff
@@ -9,243 +9,261 @@
+                'datasourceUrn': datasource_urn,
+                'description': description
+            })
 
-        if description:
-            payload['description'] = description
+    def get_dataset(self, dataset_urn, namespace_name=None):
+        if not dataset_urn:
+            raise ValueError('dataset_urn must not be None')
 
-        return self.post_request(path, path_args, payload)
-
-    @handle_response
-    def get_dataset(self, urn, namespace_name=None):
-        path = "namespaces/{0}/datasets/{1}"
         if not namespace_name:
             namespace_name = self._namespace_name
 
-        path_args = [namespace_name, urn]
-        return self.get_request(path, path_args)
+        return self._get(self._url('/namespaces/{0}/datasets/{1}', namespace_name, dataset_urn))
 
-    @handle_response
-    def list_datasets(self, namespace_name=None):
-        path = "namespaces/{0}/datasets"
+    def list_datasets(self, namespace_name=None, limit=None, offset=None):
         if not namespace_name:
             namespace_name = self._namespace_name
 
-        path_args = [namespace_name]
-        return self.get_request(path, path_args)
-
-    def put_request(self, path, path_args=None, args=None):
-        path = _compose_path(path, path_args)
-        full_path = "{0}/{1}".format(self._api_base, path)
-        return requests.put(url=full_path, json=args, timeout=self._timeout)
-
-    def post_request(self, path, path_args=None, payload=None):
-        path = _compose_path(path, path_args)
-        full_path = "{0}/{1}".format(self._api_base, path)
-
-        return requests.post(
-            url=full_path, json=payload, timeout=self._timeout)
-
-    def get_request(self, path, path_args=None, request_args=None):
-        if request_args and not isinstance(request_args, collections.Mapping):
-            raise Exception("Request argument must be a dictionary")
-        path = _compose_path(path, path_args)
-        full_path = "{0}/{1}".format(self._api_base, path)
-        if request_args:
-            arg_str = urllib.parse.urlencode(request_args)
-            full_path = "{0}?{1}".format(full_path, arg_str)
-        return requests.get(full_path, timeout=self._timeout)
-
-    @staticmethod
-    def _set_timeout(t):
-        timeout_milliseconds = (float(t) if (t and t.isnumeric())
-                                else _DEFAULT_TIMEOUT_MS)
-        return timeout_milliseconds / 1000.0
+        return self._get(
+            self._url('/namespaces/{0}/datasets', namespace_name),
+            params={
+                'limit': limit,
+                'offset': offset
+            })
+
+    def _url(self, path, *args):
+        return f'{self._api_base}{path.format(*args)}'
+
+    def _post(self, url, payload, as_json=True):
+        now_ms = self._now_ms()
+
+        response = requests.post(
+            url=url, headers=_HEADERS, json=payload, timeout=self._timeout)
+        log.info(f'{url}', method='POST', payload=json.dumps(
```

Comment:
If `requests.post` throws an exception, this won't get logged in the next line. Would this request be logged in the code that handles that exception? If not, maybe we can re-order the log entry to be before the request.

Reply:
We'll want to wrap the `requests.post`, etc in a **try/catch** to handle `ConnectionErrors`, etc see [Errors and Exceptions](https://2.python-requests.org//en/latest/user/quickstart/#errors-and-exceptions). see issue #56

>  If not, maybe we can re-order the log entry to be before the request.

The log captures the request duration, so we can't reorder. `requests` already logs a great deal allowing us to set the log level to `DEBUG` if we wanted to be more granular logs on the request.